internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -279,62 +279,80 @@ func (a *Adapter) EnsureIntegrationPipelineRunsExist() (controller.OperationResu
 
-	for _, snapshotComponent := range a.snapshot.Spec.Components {
-		if snapshotComponent.Name == componentToUpdate.Name {
-			patch := client.MergeFrom(componentToUpdate.DeepCopy())
-			componentToUpdate.Spec.ContainerImage = snapshotComponent.ContainerImage
-			err := a.client.Patch(a.context, componentToUpdate, patch)
-			if err != nil {
-				a.logger.Error(err, "Failed to update .Spec.ContainerImage of Global Candidate for the Component",
-					"component.Name", componentToUpdate.Name)
-				return controller.RequeueWithError(err)
+	// get component from component snapshot
+	if gitops.IsComponentSnapshot(a.snapshot) {
+		err = retry.OnError(retry.DefaultRetry, func(_ error) bool { return true }, func() error {
+			componentToUpdate, err = a.loader.GetComponentFromSnapshot(a.context, a.client, a.snapshot)
+			return err
+		})
+		if err != nil {
+			_, loaderError := h.HandleLoaderError(a.logger, err, fmt.Sprintf("Component or '%s' label", tekton.ComponentNameLabel), "Snapshot")
+			if loaderError != nil {
+				return controller.RequeueWithError(loaderError)
 			}
-			a.logger.LogAuditEvent("Updated .Spec.ContainerImage of Global Candidate for the Component",
-				componentToUpdate, h.LogActionUpdate,
-				"containerImage", snapshotComponent.ContainerImage)
-			if reflect.ValueOf(snapshotComponent.Source).IsValid() && snapshotComponent.Source.GitSource != nil && snapshotComponent.Source.GitSource.Revision != "" {
-				patch := client.MergeFrom(componentToUpdate.DeepCopy())
-				componentToUpdate.Status.LastBuiltCommit = snapshotComponent.Source.GitSource.Revision
-				err = a.client.Status().Patch(a.context, componentToUpdate, patch)
+			return controller.ContinueProcessing()
+		}
+
+		// look for the expected snapshotComponnet and update
+		for _, snapshotComponent := range a.snapshot.Spec.Components {
+			snapshotComponent := snapshotComponent //G601
+			if snapshotComponent.Name == componentToUpdate.Name {
+				//update .Spec.ContainerImage for the component included in component snapshot
+				err = a.updateComponentContainerImage(a.context, a.client, componentToUpdate, &snapshotComponent)
+				if err != nil {
+					return controller.RequeueWithError(err)
+				}
+
+				// update .Status.LastBuiltCommit for the component included in component snapshot
+				err = a.updateComponentSource(a.context, a.client, componentToUpdate, &snapshotComponent)
 				if err != nil {
-					a.logger.Error(err, "Failed to update .Status.LastBuiltCommit of Global Candidate for the Component",
-						"component.Name", componentToUpdate.Name)
 					return controller.RequeueWithError(err)
 				}
-				a.logger.LogAuditEvent("Updated .Status.LastBuiltCommit of Global Candidate for the Component",
-					componentToUpdate, h.LogActionUpdate,
-					"lastBuildCommit", componentToUpdate.Status.LastBuiltCommit)
+				// Updated the component for component snapshot, break
+				break
+			} else {
+				// expected component not found, continue
+				continue
+			}
+		}
+	} else {
```

Comment:
Maybe we can add a check to ensure that we only update all components for override Snapshots just in case?
Something like:
```
} else if gitops.IsOverrideSnapshot(a.snapshot) {
```

Reply:
Updated. Thanks.